backend/store/sessions.go

```diff
@@ -1,15 +1,58 @@
 package store
 
-import "github.com/highlight-run/highlight/backend/model"
+import (
+	"context"
+	"fmt"
+	"github.com/go-redis/cache/v8"
+	"github.com/highlight-run/highlight/backend/model"
+	"github.com/highlight-run/highlight/backend/redis"
+	log "github.com/sirupsen/logrus"
+	"time"
+)
 
-func (store *Store) GetSession(sessionID int) (model.Session, error) {
-	var session model.Session
+func cachedEval[T any](ctx context.Context, redis *redis.Client, cacheKey string, lockTimeout, cacheExpiration time.Duration, fn func() (*T, error)) (value *T, err error) {
+	// wait here to check the cache in case another process is waiting for db query
+	if acquired := redis.AcquireLock(ctx, cacheKey, lockTimeout); acquired {
+		defer func() {
+			if err := redis.ReleaseLock(ctx, cacheKey); err != nil {
+				log.WithContext(ctx).WithError(err).Error("failed to release lock")
+			}
+		}()
+	}
 
-	err := store.db.Where(&model.Session{
-		Model: model.Model{
-			ID: sessionID,
-		},
-	}).Take(&session).Error
+	if err = redis.Cache.Get(ctx, cacheKey, &value); err != nil {
+		if value, err = fn(); value == nil || err != nil {
+			return
+		}
+		if err = redis.Cache.Set(&cache.Item{
+			Ctx:   ctx,
+			Key:   cacheKey,
+			Value: &value,
+			TTL:   cacheExpiration,
+		}); err != nil {
+			return
+		}
+	}
 
-	return session, err
+	return
+}
+
+func (store *Store) GetSessionFromSecureID(ctx context.Context, secureID string) (*model.Session, error) {
+	return cachedEval(ctx, store.redis, fmt.Sprintf("session-secure-%s", secureID), 150*time.Millisecond, time.Second, func() (*model.Session, error) {
```

Comment:
I like this, and just want to make sure my understanding here is correct

We try the cache, if we get a hit return that value. If miss or timeout trying to get lock, we execute the code below?

Reply:
yup exactly! adding docs to the `cachedEval` function.